### PR TITLE
feat(ngmesh2obj): produce one file per mesh

### PR DESCRIPTION
**Pull request by william-silversmith**
Now produces a unifed mesh file instead of lots of mesh fragment files.
Also a major performance boost.

### CODE REVIEW

**Comment by nkemnitz** [comment]
The fragment data does not contain vertex normals?

**Comment by william-silversmith** [comment]
Nah, it seems neuroglancer computes them after download. Meshlab seems to
get along as well. Would you like them to be there?

**Comment by nkemnitz** [comment]
No, it's fine. If we kept the normals from marching cubes (before simplification), then they might be useful. But since we recalculate them from the already simplified meshes, they don't provide any additional information.

File: python/neuroglancer/pipeline/scripts/ngmesh2obj.py [hunk 1]
```diff
@@ -48,57 +48,52 @@ def decode_downloaded_data(frag_datas):
...
+      'num_vertices': num_vertices, 
+      'vertices': vertices, 
+      'faces': faces
+    }
+
+def mesh_to_obj(fragment, num_prev_verticies):
+    objdata = []
+    
+    for vertex in fragment['vertices']:
+        objdata.append('v %s %s %s' % (vertex[0], vertex[1], vertex[2]))
```
Comment (nkemnitz) [inline python/neuroglancer/pipeline/scripts/ngmesh2obj.py]: Switching between 2 and 4 space indentation across the file

File: python/neuroglancer/pipeline/scripts/ngmesh2obj.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
 import gzip
```
Comment (nkemnitz) [inline python/neuroglancer/pipeline/scripts/ngmesh2obj.py]: `StringIO`, `gzip` and `neuroglancer.lib.mkdir` seem to be unused.

File: python/neuroglancer/pipeline/scripts/ngmesh2obj.py [hunk 3]
```diff
@@ -48,57 +48,52 @@ def decode_downloaded_data(frag_datas):
   return data
 
 def decode_mesh_buffer(fragment):
-  """
-  Mesh array format is:
```
Comment (william-silversmith) [inline python/neuroglancer/pipeline/scripts/ngmesh2obj.py]: See the top comment

**Comment by william-silversmith** [comment]
I agree that the script should error for these conditions. When generating objs from a few tests, there seems to be usually one fragment that's screwed up somehow. Maybe there's a bug in MeshTask. I feel like until I debug it, notification that the obj is possibly missing a piece is enough so at least you can generate meshes to look at. 

The alternative is that I fix MeshTask and rerun meshing when we find screwed up meshes.

Here's an example error I'm seeing:

```
(venv)wms:~/neuroglancer/python/neuroglancer$ python pipeline/scripts/ngmesh2obj.py gs://neuroglancer/pinky40_v11/watershed_mst_smc_sem5_remap/ 27711747
downloading 27711747...
decoding...
Unable to process fragment mesh_mip_3_err_20/27711747:0. Violation: len vertex data != 12 * num vertices
        Array Length: 1919296123, Vertex Count: 4602
      
saving to obj...
done.
```

**Comment by william-silversmith** [comment]
Found the bug. Sometimes the JSON manifest file was included in the downloaded fragment data. Obviously it's not a valid mesh. :P I added some handling code to exclude it if its accidentally included. I changed the warnings to errors now that the meshes are downloading and generating cleanly.

File: python/neuroglancer/pipeline/scripts/ngmesh2obj.py [hunk 4]
```diff
@@ -19,86 +19,91 @@
...
+    for i in xrange(0, len(vertex_data), 12):
+      x = struct.unpack("=f", vertex_data[i:i+4])[0]
+      y = struct.unpack("=f", vertex_data[i+4:i+8])[0]
+      z = struct.unpack("=f", vertex_data[i+8:i+12])[0]
+      vertices.append((x,y,z))
+
+    faces = []
+    for i in xrange(0, len(face_data), 4):
+      vertex_number = struct.unpack("=I", face_data[i:i+4])[0]
+      if vertex_number > num_vertices:
```
Comment (nkemnitz) [inline python/neuroglancer/pipeline/scripts/ngmesh2obj.py]: should be `>=`

Comment (william-silversmith) [inline python/neuroglancer/pipeline/scripts/ngmesh2obj.py]: I thought they were numbered starting at 1?

Comment (william-silversmith) [inline python/neuroglancer/pipeline/scripts/ngmesh2obj.py]: Oh right, this is len(array) :P

**Pull request merged**